test/persistence-hooks.suite.js

```diff
@@ -575,6 +576,232 @@ module.exports = function(dataSource, should) {
+            ctx.data.extra = 'hook data';
+            next();
+          });
+
+          TestModel.create(
+            { id: 'new-id', name: 'a name' },
+            function(err, instance) {
+              if (err) return done(err);
+
+              instance.should.have.property('extra', 'hook data');
+              done();
+            });
+        });
+
+        it('triggers `persist` hook on save', function(done) {
+          TestModel.observe('persist', pushContextAndNext());
+
+          existingInstance.name = 'changed';
+          existingInstance.save(function(err, instance) {
+            if (err) return done(err);
+
+            observedContexts.should.eql(aTestModelCtx({
+              data: {
+                id: existingInstance.id,
+                name: 'changed'
+              },
+              currentInstance: {
+                id: existingInstance.id,
+                name: 'changed',
+                extra: undefined
+              },
+              options: { throws: false, validate: true }
+            }));
+
+            done();
+          });
+        });
+
+        it('applies updates from `persist` hook - on save', function(done) {
+          TestModel.observe('persist', pushContextAndNext(function(ctx){
+            ctx.data.extra = 'hook data';
+          }));
+
+          existingInstance.save(function(err, instance) {
+            if (err) return done(err);
+            instance.should.have.property('extra', 'hook data');
+            done();
+          });
+        });
+
+        it('triggers `persist` hook on update', function(done) {
+          TestModel.observe('persist', pushContextAndNext());
+
+          TestModel.update(
+            { where: { name: existingInstance.name } },
+            { name: 'changed' },
+            function(err, instance) {
+              if (err) return done(err);
+
+              observedContexts.should.eql(aTestModelCtx({
+                data: { name: 'changed' },
+                currentInstance: { name: 'changed' },
+                where: { where: { name: existingInstance.name } }
+              }));
+
+              done();
+            });
+        });
+
+        it('applies updates from `persist` hook - on update', function(done) {
+          TestModel.observe('persist', pushContextAndNext(function(ctx){
+            ctx.data.extra = 'hook data';
+          }));
+
+          TestModel.updateAll(
+            { id: existingInstance.id },
+            { name: 'changed' },
+            function(err) {
+              if (err) return done(err);
+              loadTestModel(existingInstance.id, function(err, instance) {
+                instance.should.have.property('extra', 'hook data');
+                done();
+              });
+            });
+        });
+
+        if (dataSource.connector.findOrCreate) {
+          /* TODO (pradnya), skipped because, currentInstance is not getting set
+           * correctly through findOrCreate() in dao.js */
```

Comment:
As I see it, `findOrCreate` cannot report the current instance because it may not be created yet, is my understanding correct? In that case it's ok to leave it out of the context. Just make sure you always provide `where` in the context.